reference/models.py

```diff
@@ -1,10 +1,201 @@
+        return "%s | %s" % (self.given_names, self.last_name)
+
+    def __str__(self):
+        return "%s | %s" % (self.given_names, self.last_name)
+
+    @classmethod
+    def get_or_create(cls, author, given_names, last_name, user):
+        try:
+            if author:
+                return cls.objects.get(given_names=author.given_names, last_name=author.last_name)
+            return cls.objects.get(given_names=given_names, last_name=last_name)
+        except cls.DoesNotExist:
+            author_in_reference = cls(creator = user)
+            author_in_reference.given_names = given_names
+            authauthor_in_referenceor.last_name = last_name
+            author_in_reference.save()
+            return author_in_reference
+
+
+class Reference(CommonControlField):
+    source = models.CharField(
+        _("Source"), max_length=10, blank=True, null=True
+    )
+    authors = models.ManyToManyField(
+        AuthorInReference, verbose_name=_("Authors"), blank=True
+    )
+    volume = models.CharField(
+        _("Volume"), max_length=10, blank=True, null=True
+    )
+    issue = models.CharField(
+        _("Issue"), max_length=10, blank=True, null=True
+    )
+    fpage = models.CharField(
+        _("Fpage"), max_length=10, blank=True, null=True
+    )
+    lpage = models.CharField(
+        _("Lpage"), max_length=10, blank=True, null=True
+    )
+    elocation_id = models.CharField(
+        _("Elocation Id"), max_length=10, blank=True, null=True
+    )
+    doi = models.TextField(_("DOI"), blank=True, null=True)
+    year = models.CharField(
+        _("Year"), max_length=4, blank=True, null=True
+    )
+    article_title = RichTextField(_("Article title"), null=True, blank=True)
+    citation_ids = models.ManyToManyField(
+        PubId, verbose_name=_("Pub Id"), blank=True
+    )
+    
+    panels = [
+        FieldPanel("source"),
+        AutocompletePanel("authors"),
+        FieldPanel("volume"),
+        FieldPanel("issue"),
+        FieldPanel("fpage"),
+        FieldPanel("lpage"),
+        FieldPanel("elocation_id"),
+        FieldPanel("doi"),
+        FieldPanel("year"),
+        FieldPanel("article_title"),
+        AutocompletePanel("citation_ids"),        
+    ]
+
+    def __unicode__(self):
+        return "%s | %s | %s | %s | %s" % (self.source, self.year, self.volume, self.issue, self.fpage)
+
+    def __str__(self):
+        return "%s | %s | %s | %s | %s" % (self.source, self.year, self.volume, self.issue, self.fpage)
+
+    @classmethod
+    def get(
+        cls,
+        source,
+        authors,
+        article_title,
+        year,
+        doi,
+        citation_ids,
+        volume,
+        issue,
+        fpage,
+        elocation_id
+    ):
+        q = None
+
+        for item in citation_ids:
+            if q is None:
+                q = Q(citation_ids__pub_id_value=item.pub_id_value)
+            else:
+                q = q & Q(citation_ids__pub_id_value=item.pub_id_value)
+
```

Comment:
from functools import reduce
from operator import and_
q_list = []
for item in citations_ids:
    q_list.append(Q(citation_ids__pub_id_value=item.pub_id_value))
q = reduce(and_, q_list) if q_list else None